Add extensibility point to replace the grain activator

jdom opened the pull request:
This will instantiate grains preserving the current behavior (which is using `ActivatorUtilities`, similar to MVC
s `DefaultControllerActivator`), but allows replacing this service with a custom activator.
This enables less constraining integration with some 3rd party containers, or even manually instantiating grains.

In the future, I envision the `IGrainActivationContext` instance be resolved as a scoped service as defined in #2856.

jdom commented:
cc @Eldar1205 
Eldar, would this be a suitable extension point for you? For usage, check the new tests

jdom commented:
@dotnet-bot test netfx-functional
Reason: unrelated failure: https://ci.dot.net/job/dotnet_orleans/job/master/job/netfx-functional_prtest/236/testReport/junit/NonSilo.Tests.UnitTests.SerializerTests/MessageSerializerTests/MessageTest_TtlUpdatedOnSerialization/

Eldar1205 commented:
Thanks @jdom, this will work for me, some questions/comments:

1. I think the name IGrainActivator is misleading, grain activation is an Orleans term much more broad than grain instance resolution. I suggest the name IGrainResolutionBehavior or some other name not mixed with existing Orleans concepts.

2. Why does the IGrainActivator receive an object and not a Grain? I find it better for compile time safety since all grains should inherit Grain. In particular it's important to distinguish between grain references on Orleans client side and actual Grain instances on Orleans cluster side.

ReubenBond commented on src/OrleansRuntime/Catalog/DefaultGrainActivator.cs:
```diff
@@ -0,0 +1,62 @@
+    /// <summary>
+    /// <see cref="IGrainActivator"/> that uses type activation to create grains.
+    /// </summary>
+    public class DefaultGrainActivator : IGrainActivator
```
I was about to suggest that maybe `IGrainFactory` was a more suitable name. Nevermind... 😕 would `IGrainCreator` make sense, since this doesn't actually call the activation lifecycle method on the grain?

jdom replied:
GrainFacory (or GrainActivationFactory) would be more appropriate to the abstraction that will replace our current GrainCreator, as it matches [DefaultControllerFactory](https://github.com/aspnet/Mvc/blob/760c8f38678118734399c58c2dac981ea6e47046/src/Microsoft.AspNetCore.Mvc.Core/Controllers/DefaultControllerFactory.cs), which in our case can have stuff like the facet system we are planning in the future (MVC does a VERY similar thing to our facets).
Maybe `IGrainActivationCreator`? I don't know, none seem right, but at least IGrainActivator matches MVC naming, so we at least had familiarity on our side, but I'm definitely open to change this

ReubenBond commented on src/OrleansRuntime/Catalog/GrainCreator.cs:
```diff
@@ -109,5 +96,10 @@ public Grain CreateGrainInstance(Type grainType, IGrainIdentity identity, Type s
             ((ILogConsistentGrain)grain).InstallAdaptor(factory, state, grainType.FullName, storageProvider, svc);
         }
+
+        public virtual void Release(IGrainActivationContext context, object grain)
```
Does this need to be virtual? Seems like a leftover from a previous revision.

jdom replied:
This was by design, since I expect most custom implementations will inherit from the default one and call the base in several occasions, similar to the one in our tests. This copies the DefaultControllerActivator from MVC

ReubenBond replied:
GrainCreator is internal, though

jdom replied:
oh yeah, I thought this comment was on DefaultGrainActivator, my bad. Fixed

ReubenBond commented on src/OrleansRuntime/Catalog/IGrainActivationContext.cs:
```diff
@@ -0,0 +1,20 @@
+        IGrainIdentity GrainIdentity { get; }
+
+        /// <summary>Gets the <see cref="IServiceProvider"/> that provides access to the grain's service container.</summary>
+        IServiceProvider GrainServices { get; }
```
Potential confusion between this name and the feature known as GrainServices

jdom replied:
Hmmm, can't think of a better name that would stay within the familiarity with MVC naming conventions. Perhaps `GrainActivationServices` (which is correct, I just didn't use it because it felt too verbose)?

attilah replied:
I just not fond of loosing the "Provider" suffix, but as we talked ApplicationServices maybe a better name?

jdom replied:
well, that's the point, it's not the same as ApplicationServices (actually it currently is the same instance, but this will definitely change once we add scoped services in #2856). This will have the IServiceProvider scoped to this particular grain activation, similar to MVC's context.RequestServices, which is a scoped instance separate from their ApplicationServices.

jason-bragg commented on src/OrleansRuntime/Catalog/DefaultGrainActivator.cs:
```diff
@@ -0,0 +1,62 @@
+    /// </summary>
+    public class DefaultGrainActivator : IGrainActivator
+    {
+        private readonly Func<Type, ObjectFactory> createFactory = type => ActivatorUtilities.CreateFactory(type, Type.EmptyTypes);
```
Isn't this all that is needed?  A hook to create the grain factory?

Func<Type, object[] parameters, ObjectFactory>?

Get this one thing from DI, and nothing else needs to change, right?

jdom replied:
Unfortunately not, since the need is to also perform some cleanup tasks after the instance is done with. Also, this was part of the work started in #1901 where we wanted to emulate MVC's controller activation strategy, which to be honest, make a lot of sense for us too.
This matches 1 to 1 with [DefaultControllerActivator](https://github.com/aspnet/Mvc/blob/760c8f38678118734399c58c2dac981ea6e47046/src/Microsoft.AspNetCore.Mvc.Core/Controllers/DefaultControllerActivator.cs) (except that they have an extra layer of abstraction for ITypeActivatorCache which I didn't add here, as it doesn't add any value yet)

jason-bragg replied:
Concerns:
Insufficient design - Grain lifecycle needs work, I think we all agree, but it's also very important to get it right.  I am of the opinion that upfront design is necessary.  I do not suspect that it is sufficient or desirable to simply apply the mvc controller pattern to this because they look similar.
I advocate that we do the design work necessary to address grain lifecycle needs, then, after due diligence, if the resulting design is comparable to MVC's controller pattern, then we apply it as it is more familiar to users.  This ensures we're applying the right pattern not forcing the solution into a pattern that is already familiar.
Churn - The issue that Eldar is experiencing is in part due to the fact that we've been working on this system piecemeal, creating churn.  I advocate up front design to make fewer and more complete changes with less churn.  Making a minimal change to address the previous braking change is churn, but not as much as performing this change and the risk of further changes due to lack of upfront design.
Risk - We want to get this in for 1.5, which is already in beta, which means this change will not get as much community testing as the rest of 1.5, and this is critical code.  A minimal change is far less risky.

These are my concerns, however I will happily defer to the teams decision if most of those interested in this change do not share my concerns.

jdom commented:
@Eldar1205 
1. re: naming: touched on that on another comment. I agree naming is not ideal, but stated why I still find it preferable to some other options. Nevertheless I'm open to change it if we can find a better one. Let's continue that discussion on the other comment to unify threads.
2. The reason for `object` instead of `Grain` is that it does leave the door open to have grains that do not inherit from Grain in the future, but at the same time there is not a lot that is gained from typing it to Grain anyway. Also, it matches the MVC extensibility point called IControllerActivator.

Eldar1205 commented:
Hi,

Regarding object being open to not have Grain as base class, you can always
change to object later. Until then I think the type safety gained using
Grain will benefit coherence and testability for both application code and
framework code.
Regarding interface name, aligning with MVC shouldnt be on expanse of
Orleans terminilogy consistency, and the term grain activation is very
consistent so far. Same for grain factory. I like the
IGrainActivationCreator or even more plain simple IGrainInstanceCreator.

On May 9, 2017 9:11 PM, "Julian Dominguez" <notifications@github.com> wrote:

> *@jdom* commented on this pull request.
> ------------------------------
>
> In src/OrleansRuntime/Catalog/IGrainActivationContext.cs
> <https://github.com/dotnet/orleans/pull/3002#discussion_r115561176>:
>
> > +
> +namespace Orleans.Runtime
> +{
> +    /// <summary>
> +    /// The context associated with the current grain activation.
> +    /// </summary>
> +    public interface IGrainActivationContext
> +    {
> +        /// <summary>Gets the .NET type of the grain activation instance.</summary>
> +        Type GrainType { get; }
> +
> +        /// <summary>Gets the identity of the grain activation.</summary>
> +        IGrainIdentity GrainIdentity { get; }
> +
> +        /// <summary>Gets the <see cref="IServiceProvider"/> that provides access to the grain's service container.</summary>
> +        IServiceProvider GrainServices { get; }
>
> well, that's the point, it's not the same as ApplicationServices (actually
> it currently is the same instance, but this will definitely change once we
> add scoped services in #2856 <https://github.com/dotnet/orleans/pull/2856>).
> This will have the IServiceProvider scoped to this particular grain
> activation, similar to MVC's context.RequestServices, which is a scoped
> instance separate from their ApplicationServices.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dotnet/orleans/pull/3002#discussion_r115561176>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFBYtzRF50iENs28OhGXJ0g-1im8NkCiks5r4KxFgaJpZM4NUhhe>
> .
>

jdom commented:
@dotnet-bot test netfx-functional

jdom commented:
I split the PR into 2 commits (to be merged without squashing), so that the Release part that disposes grain instances is separate, and we could in theory pull just the creation part into 1.4.2 without any behavioral change in comparison to 1.4.1.

I also renamed `IGrainActivationContext.GrainServices` to `IGrainActivationContext.ActivationServices` to make it less ambiguous with the GrainServices feature.

I did not rename `IGrainActivator`, since we spent a long time with Reuben and couldn't come up with a better name. We considered `IGrainActivationActivator` (this one would have full parity with MVC), `IGrainActivationCreator`, `IGrainResolutionBehavior`. If you have a better suggestion, please include it, otherwise we can rename later (which is unfortunate, but practical).

jason-bragg commented:
@jdom prototyping addressed most of my lingering concerns.

ReubenBond commented:
@dotnet-bot test this please

ReubenBond commented:
![image](https://cloud.githubusercontent.com/assets/203839/26502358/692634a2-41f1-11e7-8290-6735c1cff855.png)
Not sure why CI is failing here

jdom commented:
@dotnet-bot test netfx-bvt
The java stacktrace is because Jenkins was restarted

jdom commented:
@dotnet-bot test this please
Reason: unrelated failures. Now CI should be ready to run without non-deterministic failures
